lib/lf/uscalfe/fe_space_hp.h

```diff
@@ -0,0 +1,128 @@
+/**
+ * @file
+ * @brief Lagrangian finite elements of arbitrary polynomials degree
+ * @author Tobias Rohner
+ * @date   May 20202
+ * @copyright MIT License
+ */
+
+#ifndef LF_USCALFE_FE_SPACE_HP_H_
+#define LF_USCALFE_FE_SPACE_HP_H_
+
+#include <lf/mesh/utils/all_codim_mesh_data_set.h>
+#include "hp_fe.h"
+#include "scalar_fe_space.h"
+
+namespace lf::uscalfe {
+
+/**
+ * @brief Lagrangian Finite Element Space of arbitrary degree
+ */
+template <typename SCALAR>
+class FeSpaceHP : public ScalarFESpace<SCALAR> {
+ public:
+  using Scalar = SCALAR;
+
+  FeSpaceHP() = delete;
+  FeSpaceHP(const FeSpaceHP &) = delete;
+  FeSpaceHP(FeSpaceHP &&) noexcept = default;
+  FeSpaceHP &operator=(const FeSpaceHP &) = delete;
+  FeSpaceHP &operator=(FeSpaceHP &&) noexcept = default;
+
+  /**
+   * @brief Constructor: Sets up the dof handler
+   * @param mesh_p A shared pointer to the underlying mesh (immutable)
+   * @param N The polynomial degree of the Finite Element Space
+   */
+  explicit FeSpaceHP(std::shared_ptr<const lf::mesh::Mesh> mesh_p, unsigned N)
+      : ScalarFESpace<SCALAR>(std::move(mesh_p)),
+        ref_el_(mesh_p, nullptr),
+        degree_(N) {
+    init();
+  }
+
+  using ScalarFESpace<SCALAR>::Mesh;
+
+  /** @brief access to associated local-to-global map
+   * @return a reference to the lf::assemble::DofHandler object (immutable)
+   */
+  [[nodiscard]] const lf::assemble::DofHandler &LocGlobMap() const override {
+    LF_VERIFY_MSG(Mesh() != nullptr, "No valid FE space object: no mesh");
+    LF_VERIFY_MSG(dofh_p_ != nullptr,
+                  "No valid FE space object: no dof handler");
+    return *dofh_p_;
+  }
+
+  /** @brief access to shape function layout for cells
+   * @copydoc SclarFESpace::ShapeFunctionLayout(const lf::mesh::Entity&)
+   */
+  [[nodiscard]] std::shared_ptr<const ScalarReferenceFiniteElement<SCALAR>>
+  ShapeFunctionLayout(const lf::mesh::Entity &entity) const override {
+    return ref_el_(entity);
+  }
+
+  /** @brief number of _interior_ shape functions associated to entities of
+   * various types
+   */
+  [[nodiscard]] size_type NumRefShapeFunctions(
+      const lf::mesh::Entity &entity) const override {
+    return ShapeFunctionLayout(entity)->NumRefShapeFunctions();
+  }
+
+  ~FeSpaceHP() override = default;
+
+ private:
+  // R.H.: Should this really be shared pointers ?
+  lf::mesh::utils::AllCodimMeshDataSet<
+      std::shared_ptr<const lf::uscalfe::ScalarReferenceFiniteElement<SCALAR>>>
+      ref_el_;
+  std::unique_ptr<lf::assemble::UniformFEDofHandler> dofh_p_;
+  unsigned degree_;
+
+  // R.H. Detailed comment needed for this function
```

Comment:
Many more inline comments are needed to elucidate the workings of this function.